packages/graphql-server/src/plugins/useRedwoodDirective.ts

```diff
@@ -91,79 +95,129 @@ export function hasDirective(info: GraphQLResolveInfo): boolean {
 }
 
 export function getDirectiveByName(
-  info: GraphQLResolveInfo,
-  name: string
+  fieldConfig: GraphQLFieldConfig<any, any, any>,
+  directiveName: string
 ): null | DirectiveNode {
-  try {
-    const { parentType, fieldName, schema } = info
-    const schemaType = schema.getType(parentType.name) as GraphQLObjectType
-    const field = schemaType.getFields()[fieldName]
-    const astNode = field.astNode
-    const associatedDirective = astNode?.directives?.find(
-      (directive) => directive.name.value === name
-    )
+  const associatedDirective = fieldConfig.astNode?.directives?.find(
+    (directive) => directive.name.value === directiveName
+  )
+  return associatedDirective ?? null
+}
 
-    return associatedDirective || null
-  } catch (error) {
-    console.error(error)
-    return null
-  }
+export function isPromise(value: any): value is Promise<unknown> {
+  return typeof value?.then === 'function'
 }
 
-export const useRedwoodDirective = (
+function wrapAffectedResolvers(
+  schema: GraphQLSchema,
   options: DirectivePluginOptions
-): Plugin<{
-  onResolverCalled: ValidatorDirectiveFunc | TransformerDirectiveFunc
-}> => {
-  return {
-    async onResolverCalled({ args, root, context, info }) {
-      const directiveNode = getDirectiveByName(info, options.name)
+): GraphQLSchema {
+  return mapSchema(schema, {
+    [MapperKind.OBJECT_FIELD](fieldConfig, _, __, schema) {
+      const directiveNode = getDirectiveByName(fieldConfig, options.name)
       const directive = directiveNode
-        ? info.schema.getDirective(directiveNode.name.value)
+        ? schema.getDirective(directiveNode.name.value)
         : null
-
       if (directiveNode && directive) {
         const directiveArgs =
-          getDirectiveValues(
-            directive,
-            { directives: [directiveNode] },
-            info.variableValues
-          ) || {}
-
+          getDirectiveValues(directive, { directives: [directiveNode] }) || {}
+        const originalResolve = fieldConfig.resolve ?? defaultFieldResolver
         if (_isValidator(options)) {
-          await options.onResolverCalled({
-            root,
-            args,
-            context,
-            info,
-            directiveNode,
-            directiveArgs,
-          })
+          return {
+            ...fieldConfig,
+            resolve: function useRedwoodDirectiveValidatorResolver(
+              root,
+              args,
+              context,
+              info
+            ) {
+              const result = options.onResolverCalled({
+                root,
+                args,
+                context,
+                info,
+                directiveNode,
+                directiveArgs,
+              })
+
+              if (isPromise(result)) {
+                return result.then(() =>
+                  originalResolve(root, args, context, info)
+                )
```

Comment:
This is a small performance optimization for avoiding making stuff async where not needed